File: src/19_double_bracket_sequence_check/double_bracket_sequence_check.cpp
#include <iostream>
#include <string>

using namespace std;

bool valid(string s) {
  string stack;
  for (auto c : s) {
    if (c == '(' || c == '[') {
      stack += c;
      continue;
    }
    if (stack.empty()) return false;
    char need = (c == ')') ? '(' : '[';
    if (stack.back() != need) return false;
    stack.pop_back();
  }
  return stack.empty();
}

int main(void) {
  string s;
  cin >> s;
  cout << (valid(s) ? "YES" : "NO") << endl;
  return 0;
}

File: src/19_double_bracket_sequence_check/double_bracket_sequence_check.py
def valid(s):
  stack = []
  for c in s:
    if c in ('(', '['):
      stack.append(c)
      continue
    if not stack:
      return False
    need = '(' if c == ')' else '['
    if stack.pop() != need:
      return False
  return not stack

s = input()
print('YES' if valid(s) else 'NO')
